arangod/Aql/OptimizerRules.cpp

```diff
@@ -822,6 +822,15 @@ std::string getSingleShardId(
 
   // all shard keys found!!
 
+  if (node->getType() == EN::INSERT && collection->numberOfShards() != 1 &&
```

Comment:
I was not quickly able to decide why this if block has to be here. If `shardKeys` contains at least one element which does not contain a dot, then `toFind` is not empty, and so we never reach this if. Empty shardKeys is not allowed (as far as I know), and so the only case in which this would be triggered would be if shardKeys has one or more entries containing a dot, which is also not allowed (according to some comment above).

Reply:
`toFind` is initially populated, but some code between its population and line 825 can remove items from `toFind`. 
If `toFind` is empty, it actually means that the lookup conditions of the query covered all the shard keys of the collection, so in case we get past the `if (!toFind.empty()) { return ... }` it simply means that we know the lookup values for the collection's shard keys.

Reply:
OK, understood, thanks for the explanation!